### DISCUSSION TITLE
Feature Request: Array destructuring

### DISCUSSION THREAD

**Discussion by Meir017** [dotnet/csharplang#388] [General]
I would like to be able to destruct an array into multiple objects similar to js's feature
````js
var array = [1,2,3];
var [ b, c, d] = array;
// b = 1, c = 2, d = 3
````

The syntax could be similar to tuple destructuring:
````csharp
int[] array = { 1, 2, 3 };

int [ a, b, c] = array;
// or
[int a, int b, int c] = array;
// or
var [ a, b, c] = array;
````

**Comment by stepanbenes** [comment]
I think it is already possible in C#7 to use tuple destructuring syntax applied on array by using user-defined extension method `Deconstruct`.

**Comment by HaloFour** [comment]
```cs
public static class ArrayDestructuringExtensions {
    public static void Deconstruct<T>(this T[] array, out T item1, out T item2, out T item3) {
        item1 = default(T);
        item2 = default(T);
        item3 = default(T);
        if (array != null) {
            if (array.Length > 2) {
                item3 = array[2];
            }
            if (array.Length > 1) {
                item3 = array[1];
            }
            if (array.Length > 0) {
                item3 = array[0];
            }
        }
    }
}
```

What's kludgy in C# 7.0 is that deconstruction is guaranteed to succeed, so if you try to deconstruct an array that is too small your options are to throw or to succeed with some default value (I chose the latter above).  With conditional patterns you should have more options.

Oh, and you'd need an overload per arity.  Here's hoping for array patterns.

**Comment by jnm2** [comment]
Why did you close this? I don't think there's a good solution yet and I've already been in situations where I could use one.

**Comment by Meir017** [comment]
https://github.com/dotnet/csharplang/issues/388#issuecomment-290210793 is the solution

**Comment by jnm2** [comment]
@HaloFour included the deficiencies of the approach in the same comment.

**Comment by aluanhaddad** [comment]
> What's kludgy in C# 7.0 is that deconstruction is guaranteed to succeed, so if you try to deconstruct an array that is too small your options are to throw or to succeed with some default value (I chose the latter above). With conditional patterns you should have more options.

How did this happen? This needs fixing. Deconstruction is almost an absurd feature without conditionality.

**Comment by HaloFour** [comment]
@aluanhaddad 

That would be #45.  Conditional deconstruction would be a part of the wider feature of pattern matching which was deferred for more bake time.

**Comment by riddlemd** [comment]
Would love to see this moved forward a bit... seems silly it died on the vine in 2017...

**Comment by sgtrusty** [comment]
As a (mostly but newly) JavaScript developer, I would like to use this when using C#. Thanks 😊.

**Comment by 333fred** [comment]
This could be done as an extension to https://github.com/dotnet/csharplang/issues/3435. Specifically, it would be filling in this chart: https://github.com/dotnet/csharplang/issues/3107.

**Comment by GF-Huang** [comment]
We need something like this:
```cs
var (host, port) = "github.com:443".Split(':');
```

**Reply by huoyaoyuan** [reply]
This is covered by list pattern, which has started to be implemented:
```csharp

if ("github.com:443".Split(':') is [var host, var port])
```

**Reply by 333fred** [reply]
No, `if ("github.com:443".Split(':') is { var host, var port })`

**Reply by GF-Huang** [reply]
Is this works on c# 9 now?

**Reply by 333fred** [reply]
No. This is a proposal that might make it in for 10: https://github.com/dotnet/csharplang/issues/3435.

**Reply by Tobberoth** [reply]
For anyone showing up later, this was implemented in C# 11 and huoyaoyuan got the syntax correct.

**Reply by 333fred** [reply]
Yes. At the time, we were still hoping to be able to use the curly braces, but we switched to square brackets later.

**Reply by jeremysimmons** [reply]
>For anyone showing up later, this was implemented in C# 11 and huoyaoyuan got the syntax correct.

@Tobberoth link?

**Reply by CyrusNajmabadi** [reply]
@jeremysimmons

```c#
if ("github.com:443".Split(':') is [var host, var port])
```

**Reply by GF-Huang** [reply]
> @jeremysimmons
> 
> ```cs
> if ("github.com:443".Split(':') is [var host, var port])
> ```

Is this supports in .NET 7?
